src/sentry/snuba/metrics/metric_extraction.py

```diff
@@ -0,0 +1,258 @@
+    exclusive, behavior is undefined if both are specified.
+    """
+
+    key: str
+    field: NotRequired[str]
+    value: NotRequired[str]
+    condition: NotRequired[RuleCondition]
+
+
+class MetricSpec(TypedDict):
+    """
+    Specification for a metric to extract from some data.
+
+    The metric type is given as part of the MRI (metric reference identifier)
+    which must follow the form: `<type>:<namespace>/<name>@<unit>`.
+
+    How the metric's value is obtained depends on the metric type:
+     - Counter metrics are a special case, since the default product counters do
+       not count any specific field but rather the occurrence of the event. As
+       such, there is no value expression, and the field is set to `None`.
+       Semantics of specifying remain undefined at this point.
+     - Distribution metrics require a numeric value.
+     - Set metrics require a string value, which is then emitted into the set as
+       unique value. Insertion of numbers and other types is undefined.
+    """
+
+    category: Literal["transaction"]
+    mri: str
+    field: NotRequired[Optional[str]]
+    condition: NotRequired[RuleCondition]
+    tags: NotRequired[Sequence[TagSpec]]
+
+
+def convert_alert_to_metric(snuba_query: SnubaQuery) -> Optional[MetricSpec]:
+    if snuba_query.dataset != Dataset.PerformanceMetrics.value:
+        return None
+
+    try:
+        spec = OndemandMetricSpec.parse(snuba_query.aggregate, snuba_query.query)
+        if not spec:
+            return None
+
+        return {
+            "category": DataCategory.TRANSACTION.api_name(),
+            "mri": spec.mri,
+            "field": spec.field,
+            "condition": spec.condition(),
+            "tags": [{"key": QUERY_HASH_KEY, "value": spec.query_hash()}],
+        }
+    except Exception as e:
+        logger.error(e, exc_info=True)
+        return None
+
+
+@dataclass(frozen=True)
+class OndemandMetricSpec:
+    _field: str
+    _query: str
+
+    field: Optional[str]
+    metric_type: str
+    mri: str
+    op: MetricOperationType
+
+    @classmethod
+    def check(cls, field: str, query: str) -> bool:
+        """Returns ``True`` if a metrics query requires an on-demand metric."""
+        # TODO?: self.dataset == Dataset.PerformanceMetrics.value and
+        return "transaction.duration" in query
+
+    @classmethod
+    def parse(cls, field: str, query: str) -> Optional["OndemandMetricSpec"]:
+        if not cls.check(field, query):
+            return None
+
+        relay_field, metric_type, op = _extract_field_info(field)
+
+        return cls(
+            _field=field,
+            _query=query,
+            field=relay_field,
+            metric_type=metric_type,
+            mri=f"{metric_type}:{CUSTOM_ALERT_METRIC_NAME}@none",
+            op=op,
+        )
+
+    def query_hash(self) -> str:
+        str_to_hash = f"{self.field};{self._query}"
+        return hashlib.shake_128(bytes(str_to_hash, encoding="ascii")).hexdigest(4)
```

Comment:
I would add a comment here stating as a `TODO` that we have to clarify the spec w.r.t. ordering and how we want to deal with multi-conditions (the big problem is that we can have nested expressions, thus we would have to order the AST if we wanted to properly implement deterministic hashing).